### ISSUE TITLE
[Proposal] Actor Reentrancy

### ISSUE THREAD

**Issue by halspang** [dapr/dapr#2901]
## In what area(s)?

/area runtime
> /area operator
> /area placement
> /area docs
> /area test-and-release

### Background
The [Virtual Actor Pattern](https://www.microsoft.com/en-us/research/project/orleans-virtual-actors/) provides a simple framework for developers to create distributed applications with many of the common headaches of distributed apps handled by the framework. In simple terms, Actors provide an abstraction that allows for easy concurrency/scaling, fault tolerance, and state management. A key attribute of Actors is that they are single threaded giving in-order execution.

Information on Dapr's version of Actors can be seen [here](https://docs.dapr.io/developing-applications/building-blocks/actors/actors-overview/).

### Problem
Currently, Dapr v1.0, does not allow for Actor Reentrancy. This, in conjunction with Actors being single threaded causes problems when Actors wish to communicate with other Actors. For example, say there are 3 Actors {A, B, C}. These Actor methods are reliant on each other, which leads to a call structure such like:

Actor A -> Actor B -> Actor C -> Actor A

In Dapr v1.0, this will cause a deadlock as the call from Actor C to Actor A will wait for the original call into Actor A to finish before allowing Actor C's call. Thus, Actors currently cannot be used for more complex workflows. The solution here is to add Actor Reentrancy to Dapr's framework.

Actor Frameworks such as [Orleans](https://dotnet.github.io/orleans/) and [Service Fabric Reliable Actors](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-reliable-actors-introduction) allow for Actor Reentrancy. 

### Solution
For this problem, we'll be following the path set by Orleans/Service Fabric fairly closely. In broad strokes, what we will do is:

- Provide a config entry that specifies if an Actor is reentrant (opt-in)
- Provide a call context (actor's present in chain) to the actor framework
- If the actor being called is present in the call context, bypass the actor's lock

#### Runtime
The runtime will be responsible for handling the lock/reentry. It must be able to read the new config value (defaulting to false) and parse/understand the call context. 

We are choosing this to be an opt-in feature as dapr v1.0 does not support it so the customer must actively change their code in order to make use of this feature. This will stop unplanned changes in their functionality.

#### SDKs/Raw Actors
The SDKs and raw actors must be the ones to provide the reentrancy information to the dapr runtime. This includes building the call context and providing the new configuration value to the runtime. 

The call context should be a Set of Actors that are present in the call chain. Each time an Actor tries to call another Actor, the current Actors type and ID should be inserted into the Set.

The config value can be specified with whatever language specific options are available (like annotations) or through a raw config which is served with the request to `/dapr/config`.

@yaron2 @artursouza

**Comment by artursouza** [comment]
Can the SDK just pass through the context (via http header or grpc metadata) and let the Dapr runtime add/remove from the set of actors?

**Comment by withinboredom** [comment]
Playing devil's advocate here to locate edge/failure cases:

1. What happens if an actor is moved from a host during a call chain, will it be retried?
2. If the context is passed via headers, what happens when it grows beyond 8kb and servers start responding with a `413`?
3. What happens when this is used to fan-out and runs actors in a multi-threaded context?

I think 1. will be "easy" to address, but 2. presents some very real issues that enforce an implicit maximum stack-depth that would only apply to services running as non-GRPC and might be difficult to debug. 

It might be better to pass an opaque context to services/actors which is then looked up in the store actor store to add/remove active actors from a set. Considering that actor stores now require an etag, this should be pretty straightforward to handle fan-out.

For 3. it's where things get interesting... Imagine the following chain:

1. Actor A:123 calls B:123.
2. B:123 calls several methods of A:123 in parallel

Now A:123 is allowed to be called in parallel (the lock is bypassed) and the single-threaded guarantee of actors is violated. I'm honestly not sure how to fix it; or if it should be fixed.

**Comment by halspang** [comment]
@artursouza My thought for having the Actor provide the context is that it makes the chain to the actor runtime being distributed/failing/moving to a new host. If an Actor fails in the middle of the chain, the request is going to fail regardless, but if we called a runtime without the persisted chain we could see incorrect behavior.

@withinboredom Thanks for bringing up some more failure cases! 

1. I don't think there's a need to replace Actors in the middle of a call as I still see the entire call stack as 1 actor operation, just involving multiple actors. If the one at the very end of the chain fails, we can retry as normal, but if something in the middle fails I think the request will just fail to that point where it can then be retried or the entire request can retry depending on how the error is handled.
2. Is what you're suggesting here making it part of the request itself? I agree the 8KB limit in headers will be an issue and to some extent, Artur's suggestion solves it but I feel it creates other issues with the different actor runtimes agreeing on the context. 
3. This is an interesting scenario. There's some precedent for letting actors run in parallel as shown with Orlean's Interleaved annotation. I'm not sure if that's the right way to go here though. We could start getting on the entire chain instead of having it be a set of who's involved have it be a set of unique entry points where only 1 call is allowed per unique entry. Though I still worry about what happens if the actor runtime dies. We could always move to storing the chain in a dedicated store which might not add too much latency. That could open up some performance concerns, but might be worth investigating.

**Comment by withinboredom** [comment]
> I don't think there's a need to replace Actors in the middle of a call as I still see the entire call stack as 1 actor operation, just involving multiple actors. If the one at the very end of the chain fails, we can retry as normal, but if something in the middle fails I think the request will just fail to that point where it can then be retried or the entire request can retry depending on how the error is handled.

I think the biggest issue is that it would leave the "unknowable global application state" potentially in an undefined state. If you were to go with Artur's suggestion, then it's possible state commits could be held by daprd until the stack fully unwinds, at which point the state could be committed. That won't solve the problem of any side-effects from the calls though. But the more I think about it, I think this case is just something to be aware of when writing reentrant actors.

> Is what you're suggesting here making it part of the request itself? I agree the 8KB limit in headers will be an issue and to some extent, Artur's suggestion solves it but I feel it creates other issues with the different actor runtimes agreeing on the context.

We're already storing actor reminders in the state store. So, that could be a solution. The context could simply be the `(key, etag)` of the current call chain, so if you didn't have the right etag, you couldn't make another call. SDKs would have to accept the context from an incoming request and update it from any responses. This also handily solves the parallel case, by simply causing an error with an invalid etag once it processes the first request.

**Comment by halspang** [comment]
> I think the biggest issue is that it would leave the "unknowable global application state" potentially in an undefined state. If you were to go with Artur's suggestion, then it's possible state commits could be held by daprd until the stack fully unwinds, at which point the state could be committed. That won't solve the problem of any side-effects from the calls though. But the more I think about it, I think this case is just something to be aware of when writing reentrant actors.

Yeah, the potential unknown global state is a bit concerning. It's why I like the idea of having the actors control it as it lives and dies with the actors. As discussed though, that brings its own issues. A guide for reentrant actor development could definitely address a lot of this as most paths seem to have a case where we can lose our state. 

> We're already storing actor reminders in the state store. So, that could be a solution. The context could simply be the (key, etag) of the current call chain, so if you didn't have the right etag, you couldn't make another call. SDKs would have to accept the context from an incoming request and update it from any responses. This also handily solves the parallel case, by simply causing an error with an invalid etag once it processes the first request.

True, I am starting to like the idea of using a statestore more. I was thinking something similar with the etags acting as an additional gating mechanism. Using TTLs we could even clean up some old state if it gets orphaned for any reason. If the key is a random guid and the actor operation is idempotent it clears up a lot of issues.

**Comment by withinboredom** [comment]
> Using TTLs we could even clean up some old state

I was thinking the same thing :) 

> Yeah, the potential unknown global state is a bit concerning.

Agree, but I'm sure once we see how it all actually works and use it, it won't be as concerning as we think it currently is. Being able to query the underlying store and discover orphaned call chains would be a great boon to help locate bugs in dapr, sdks, and applications. In reality, it's no riskier than it currently is; we'll just be able to monitor it. Anything you can monitor can be improved.

Ideally, SDKs would provide a way to commit the state arbitrarily such that if you _really, really_ want to make sure that data gets stored, you can commit before calling another actor/service.

**Comment by halspang** [comment]
Sounds like we're coming to a conclusion here. I'll write up a more formal proposal of how we'll handle the state and we can take further comments from there. 

> Ideally, SDKs would provide a way to commit the state arbitrarily such that if you really, really want to make sure that data gets stored, you can commit before calling another actor/service.

That's a good idea. I feel like that's getting close to using actors as a proxy for things like Azure Logic Apps or AWS Step Functions. It'd be cool to be able to compose a workflow with actors as the building block.

**Comment by rynowak** [comment]
Coming into this a bit late but I want to try and get clarifications about a few things re: the programming experience...

I'll include some examples using C# - but my hope here is at all of the SDKs behave *similarly enough* for these principles to be true across all of them. 

## Quiescence

Do we expect reentrant calls to only occur at quiescencent points - specifically as a result of an **outgoing** call to another actor? Are there any cases where a reentrant call could be observed at another point?

Example of what I means:

```C#
public Task DoSomething()
{
    // some business logic here

    var proxy = this.ProxyFactory.CreateActorProxy<IMyOtherActor>(....);

    // begin quiescencent point here
    await proxy.InvokeSomeReentrantFunctionality(); 
    // begin quiescencent point here

   // more business logic here
}

// Called by IMyOtherActor.InvokeSomeReentrantFunctionality()
public Task MyMethodCalledReentrantlyByAnotherActor()
{
   // my logic runs while DoSomething is waiting at the 'await'
}
```

The guarantee is that since the actor instance itself always triggers reentrancy via an outgoing call, you never observe an unintended interleaving (reentrant call happening when you your business logic code is running).

Is this always the case? It has some really desirable semantics:

- There's no *true* multi-tasking here. Only one execution is making progress inside the actor instance at a time.
- From a developer POV this is most similar to garden-variety method calls, you call to the proxy ends invoking another method on your instance. Your call to the proxy does not return control until that other call returns control (and more).

----

From the SDK POV - if your language has *fire-and-forget* asynchrony ideally there should be some guardrails in the SDK to ensure that these invariants are maintained if possible.

There's an easy pattern that breaks the invariant:

```C#
public Task DoSomething()
{
    // some business logic here

    var proxy = this.ProxyFactory.CreateActorProxy<IMyOtherActor>(....);

    // start a reentrant call
    var task =  proxy.InvokeSomeReentrantFunctionality(); 

   // more business logic here

   // wait for completion of the other call
   await task
}
```

This will result in concurrent execution since you didn't wait for `InvokeSomeReentrantFunctionality` to complete before processing more business logic. This kind of pattern could be supported if its desirable, but it's significantly more complex  for users and the SDK.

## State and Transactionality

What are the expectations for state management with reentrancy?

```C#
public Task DoSomething()
{
    this.StateManager.SetState("A", 10);

    var proxy = this.ProxyFactory.CreateActorProxy<IMyOtherActor>(....);
    await proxy.InvokeSomeReentrantFunctionality(); 
 
   var b = this.StateManager.GetState<int>("B");
   this.StateManager.SetState("A", b+1);
}

// Called by IMyOtherActor.InvokeSomeReentrantFunctionality()
public Task MyMethodCalledReentrantlyByAnotherActor()
{
    var b = this.StateManager.GetState<int>("A");
    this.StateManager.SetState("B", b+1);
}
```

Should `MyMethodCalledReentrantlyByAnotherActor` be able to observe uncommitted changes made by `DoSomething`? 

If so, should committing changes made by `MyMethodCalledReentrantlyByAnotherActor` commit the changes made by `DoSomething` when it completes?

*I don't have a fully-formed opinion about this. Looking to start a discussion.*

**Comment by withinboredom** [comment]
I don't see an issue with your second case. Where I do see an issue is if you have multiple outgoing calls that you don't wait for; that should fail IMHO (can't make more than one re-entrant call at a time).

As for the state, it probably *should* be able to observe an uncommitted state, as it's the same instance; there shouldn't be more than one instance of an actor.

**Comment by halspang** [comment]
@rynowak I'm not entirely sure what you mean by other places where we can be reentrant. Are you suggesting internal calls within the actor triggering it as well? 

I like the idea of putting guardrails on the SDK, my worry here is that we run into issues where chains call back and forth into the same actors. [Orleans](https://dotnet.github.io/orleans/docs/grains/reentrancy.html#reentrancy-within-a-call-chain) has this as their example where two methods are called in a loop until a return case is hit. If we're calling into the same actor with the same context, I feel like it could cause issues. I'm definitely not against this idea, just want to figure out the impact of it.

I also don't see a problem with a specific actor instance sharing state if the SDK supports it. However, I do think we should have a consistent experience across the SDKs so that could lean the decision one way or the other.

@withinboredom If you only want 1 reentrant call at a time, are you referring to chains like A->B->A->B where technically A has made 2 reentrant calls? Or specifically not allowing A to call B more than once at a time from a specific method like Ryan mentioned where you could fire off several calls without awaiting?

**Comment by halspang** [comment]
That's what I thought you were referring to and I agree.

What I do want to keep is the `A -> B -> A -> B` chain which still feels possible to me as the 2nd A has a different context from the first so it can still fire off a new reentrant request.

**Comment by rynowak** [comment]
> I don't see an issue with your second case. Where I do see an issue is if you have multiple outgoing calls that you don't wait for; that should fail IMHO (can't make more than one re-entrant call at a time).

The issue with the second case is that it becomes a concurrent operation because two operations are making progress at a time. Libraries are either designed for concurrent access or they are not (usually not). 

In a language/library with blocking semantics this isn't possible (because blocking). In a language with async/await this is a pretty common mistake for a beginner. The runtime can't do anything to help here because it doesn't have visibility about whether the code is at a quiescencent point. The runtime *can* as @withinboredom pointed out... help here if you start a second reentrant call.

I've seen other actor frameworks make reentrancy a per-call decision rather than per-type/method for fine-grained control. 

> I'm not entirely sure what you mean by other places where we can be reentrant. Are you suggesting internal calls within the actor triggering it as well?

I'm not sure... I'm asking 😆  I wanted to make my assumptions explicit in case I was wrong.

>I also don't see a problem with a specific actor instance sharing state if the SDK supports it. However, I do think we should have a consistent experience across the SDKs so that could lean the decision one way or the other.

💯% this is where I'm at. I don't know the right semantics, but it seems like we should figure out the pattern, and it should be  should be the same across SDKs.

If it helps this discussion, I have some old sample code I wrote with .NET actors a while back https://github.com/rynowak/rock-paper-actors that I ultimately gave up on because I needed reentrancy to be successful. I could update that code and we could try to use some of those examples as motivation to get input on this case.

**Comment by withinboredom** [comment]
> The issue with the second case is that it becomes a concurrent operation because two operations are making progress at a time.

Ah, I see. If A -> A without `async`, you'd end up with two methods running at the same time; or even A -> B. I think that sounds like a legitimately hard problem. :D There's probably not much the runtime could do to prevent someone from doing that short of a custom task scheduler in C# (I have no idea how that would be implemented in other languages)?

> and it should be should be the same across SDKs.

I totally agree. With the recent optimizations I've made to actors, it looks like it's possible the PHP SDK could support either read-uncommitted or read-committed actors.

**Comment by withinboredom** [comment]
> I've seen other actor frameworks make reentrancy a per-call decision rather than per-type/method for fine-grained control.

Technically, an SDK could expose it as per-call. Just don't attach the rentrant context to the call and reentrency would be disabled for that call. Right?

> I would like to avoid writing another custom sync context

Yeah, those aren't fun... or rather, fun in the way that makes one's hair gray. 😂

**Comment by halspang** [comment]
> Technically, an SDK could expose it as per-call. Just don't attach the rentrant context to the call and reentrency would be disabled for that call. Right?

Yeah, we could also add on a reentrant header that just lets the runtime know to look for it if we wanted to be really explicit about it.

> Yep. I think the most straightforward answer is to say don't do that.

Agreed. I'm up for a number of guardrails and I think at the very least we can add in something where the runtime can recognize the context that the call is coming from and only let one request through at a time from any given context. That plus docs feels like it gets us most of the way to where we want to be but I wouldn't be surprised if these shifted over the course of the implementation/discovering new issues.

**Comment by artursouza** [comment]
I am late for the discussion going on here. Can the proposal be updated with the latest version after all the discussion?

One thing we should keep in mind is which functionality belongs to sidecar vs SDK. Historically, we push as much as possible to the sidecar. For example, it seems that SDK will pass the call context along but should the SDK handle updates to the context too? I will wait for an update to the proposal before I comment on specifics.

**Comment by halspang** [comment]
@withinboredom / @rynowak  Here's an attempt at diving into how we can handle the context. I feel pretty good about most of it, the only thing I don't really like is how I'm handling removing stuff from the chain. I'll keep thinking on it to see if there's something more elegant but I welcome feedback in the meantime.

@artursouza Here's the biggest update to the proposal thus far. I'll leave it open to comments but can put it all together in one post now if that's easier to process.

## State and the Call Context
The call context is integral to reentrancy as it lets us determine if an actor can bypass its lock or not. Thus, the call chain must be resilient to:

- 8KB header limits and request size (larger and larger requests for deep chains)
- Host death/new placement
- Conflicting requests (multi-threaded actor calls/interleaving)

Accomplishing this leads us towards using a statestore to store the actual context chain, similar to how reminders are persisted. This also simplifies the SDK code as storing the chain does not need to be replicated across all SDKs.

### Runtime Changes
The runtime is responsible for handling the state of the call context which includes generating it. It also enforces the various guarantees we're trying to make. The high level view of it is:

1. Ensure actor is reentrant (config)
2. Generate reentrant/context guids
3. Store/update/read context
4. Block reentrancy when the context doesn't meet the requirements
5. Utilize ETags to stop concurrent calls from a single actor

Attempting to write this all out more specifically made me think that pictures might do better, so here's what I imagine the overall and a zoomed in flow of the context processing would look like.

#### High level flow
<img width="630" alt="actor_runtime_reentrancy" src="https://user-images.githubusercontent.com/70976921/110996323-a253c580-8330-11eb-91ca-d0980f0801a4.png">

#### Zoom in on Chain Process
<img width="444" alt="actor_reentrancy_append_logic" src="https://user-images.githubusercontent.com/70976921/110996359-aed81e00-8330-11eb-9949-ee1dc2eb7368.png">

### SDK Changes
The changes in the SDK become simple under this method. The main responsibilities are:

- Providing a way for the config to specify reentrant
- Passing the call context id through as a header

**Comment by withinboredom** [comment]
Would it be simpler to let the SDK determine if it's reentrant or not? That would allow for SDKs to allow reentrancy on specific methods or entire actors, or not at all. We could even commandeer error 451 if reentrancy is not allowed.

**Comment by halspang** [comment]
> Would it be simpler to let the SDK determine if it's reentrant or not? That would allow for SDKs to allow reentrancy on specific methods or entire actors, or not at all. We could even commandeer error 451 if reentrancy is not allowed ( joking but not joking - in reality, any error would/should go back to the caller ).

The reason I moved a lot of the logic to the runtime was to avoid replicating functionality across the different SDKs. They already provide a config, so that seemed a good place to add the flags for if an actor was reentrant or not. 

My other concern was giving the SDKs the knowledge of where they were in a reentrant chain. If Actor A calls Actor B and Actor B's method isn't reentrant, this call should still succeed. However, if it was B -> A -> B, that call should fail. As you can see, this would require the SDKs to consume the entire context and understand who was calling who. Due to this, I like the centralized nature of passing this logic to the runtime as well as not having to replicate logic across different SDKs/languages. 

I do like the idea of being able to specify reentrancy on a method/actor level though. What are your thoughts on adding that info to the config? We could then have it be part of the checks I outlined in the diagrams above.

Another option here would be having the runtime pass a special header when a call is strictly reentrant (2nd call into the same actor). SDKs could key off of this which is simple logic so shouldn't have too high a cost to replicate. This could give you the behavior you want. 

I'm open to both ideas, but still think I prefer keeping the logic inside of the runtime.

**Comment by rynowak** [comment]
> My other concern was giving the SDKs the knowledge of where they were in a reentrant chain

The SDKs having the knowledge isn't the same as the SDK making the determination though 😆 The runtime could implement **the rules** but still pass on **the knowledge**.

We still haven't tried to answer this question:

> What are the expectations for SDK state management with reentrancy?

**Comment by halspang** [comment]
>The SDKs having the knowledge isn't the same as the SDK making the determination though 😆 The runtime could implement the rules but still pass on the knowledge.

Yup, which was what this was trying to say, is that not what you're referring to:

> Another option here would be having the runtime pass a special header when a call is strictly reentrant (2nd call into the same actor). SDKs could key off of this which is simple logic so shouldn't have too high a cost to replicate. This could give you the behavior you want.

> We still haven't tried to answer this question:

Are you asking about sharing non-committed state or their responsibility with the chain?

**Comment by artursouza** [comment]
> > The SDKs having the knowledge isn't the same as the SDK making the determination though 😆 The runtime could implement the rules but still pass on the knowledge.
> 
> Yup, which was what this was trying to say, is that not what you're referring to:

Today, the runtime is responsible for determining if an actor can be called and not the SDKs. This simplifies a lot in the SDK design since code does not need to be thread-safe, for example. Now with reentrancy, I would avoid changing this principle by adding any concurrency logic (or rules) to the SDKs. Ideally, the SDKs should not do more than pass through the context.

> 
> > Another option here would be having the runtime pass a special header when a call is strictly reentrant (2nd call into the same actor). SDKs could key off of this which is simple logic so shouldn't have too high a cost to replicate. This could give you the behavior you want.
> 
> > We still haven't tried to answer this question:
> 
> Are you asking about sharing non-committed state or their responsibility with the chain?

**Comment by rynowak** [comment]
> Are you asking about sharing non-committed state or their responsibility with the chain?

The non-committed state part. 

sorry, I think I skipped a few comments, we're on the same page now.

**Comment by halspang** [comment]
> The non-committed state part.

I think we left off saying that we just want consistent behavior between the SDKs.

Personally, I don't really like the idea of sharing non-committed state. If you want to share state between actors on a reentrant call, I think we should state you have to commit state in a place where it can be fetched. Generally, I think this is better because I think it lessens (not eliminates) the possibility of an actor leaving state in an incorrect/corrupted state if it fails part way through the method. So really this is just a gut feeling of sharing non-committed state increasing the chances of people putting themselves into bad situations. 

This is also based a bit around my thinking that an entrance to an actor method should be fairly isolated. Yes, the method can be part of a chain but if Actor A calls Actor B, B doesn't have access to uncommitted state. Calling back into A should be considered another independent event and not have access to the original, uncommitted state. I think this provides a cleaner idea of what Actors are and gives a better narrative on how to design actor workflows.

So, to be clear. My recommendation here is that we don't allow shared access to non-committed state.

I'd love to hear your opinions on this though @rynowak @artursouza @withinboredom.

**Comment by withinboredom** [comment]
I also agree that we should be consistent. Perhaps it would be a better idea for us to break off into another issue specific to the SDKs to discuss that specifically?

**Comment by artursouza** [comment]
> > The non-committed state part.
> 
> 
> 
> I think we left off saying that we just want consistent behavior between the SDKs.
> 
> 
> 
> Personally, I don't really like the idea of sharing non-committed state. If you want to share state between actors on a reentrant call, I think we should state you have to commit state in a place where it can be fetched. Generally, I think this is better because I think it lessens (not eliminates) the possibility of an actor leaving state in an incorrect/corrupted state if it fails part way through the method. So really this is just a gut feeling of sharing non-committed state increasing the chances of people putting themselves into bad situations. 
> 
> 
> 
> This is also based a bit around my thinking that an entrance to an actor method should be fairly isolated. Yes, the method can be part of a chain but if Actor A calls Actor B, B doesn't have access to uncommitted state. Calling back into A should be considered another independent event and not have access to the original, uncommitted state. I think this provides a cleaner idea of what Actors are and gives a better narrative on how to design actor workflows.
> 
> 
> 
> So, to be clear. My recommendation here is that we don't allow shared access to non-committed state.
> 
> 
> 
> I'd love to hear your opinions on this though @rynowak @artursouza @withinboredom. 

I think having access to uncommitted state makes because otherwise it would have a conflict in state on commit time. If further isolation is expected, we can discuss that as a feature addition where the state isolation is configurable.

**Comment by rynowak** [comment]
> My recommendation here is that we don't allow shared access to non-committed state.

This is my default assumption, but I don't want to rush to anything without thinking through some concrete scenarios.

**Comment by withinboredom** [comment]
Also agree. Looks like no separate issue needed.

**Comment by halspang** [comment]
Updated design:

### Decisions Made
- Launch as a preview first to confirm decisions
- Call chain/context will not be persisted but rather live in memory
- GUID will be related with lock in runtime, if lock matches the GUID we can bypass it
- State will not be shared in initial version
- Document not to use multi-threaded calls to other actors
- Impose an arbitrary limit for reentrant depth

### Future Investigations/Imrpovements
- Shareable non-committed state
- Request/response paradigm investigation
- Client side cancellation (reject further reentrant calls)
- Reentrant chain timeout
- Block multi-threaded calls

## Design
### Runtime
The runtime still assumes most of the responsibility here. The key responsibilities are:

- Recognize an Actor as reentrant (config parsing)
- Generate a GUID to associate with the Actor lock
- Match received GUID to lock's GUID to allow reentrant call

#### Configuration Changes
The configuration now needs to express reentrancy options. The modifications to the configuration are all for the one retrieved by dapr from an actor application on `dapr/config`. Here is an example modified configuration:

```json
{
  "entities":["actorType1", "actorType2"],
  "actorIdleTimeout": "1h",
  "actorScanInterval": "30s",
  "drainOngoingCallTimeout": "30s",
  "drainRebalancedActors": true,
  "reentrancy": {
    "enabled": true,
    "methods": [],
    "limit": 32
  }
}
```

Field | Required | Default | Description
-----|----------|---------|------------
reentrancy|False|Reentrancy Disabled|Tag containing reentrancy specific configuration, if not specified reentrancy is disabled.
reentrancy.enabled|True|N/A|True or false, set if reentrancy is enabled or disabled.
reentrancy.methods|False|Allow all|Filter specifying allowed methods, ignored if empty/omitted.
reentrancy.limit|False|32|Limit for the number of concurrent reentrant requests to an actor, further requests are denied.

Note: limit for reentrancy is subject to change.

#### GUID Based Locking
The core of the reentrant behavior is being able to bypass the actor locking mechanism during on ongoing request provided that it is part of the current reentrant call. To do this, the runtime will:

1. Check the request for a reentrancy ID header.
   - If there is no reentrancy ID, generate a new GUID.
1. Attempt to lock with the reentrancy ID.
   - If there is no lock, perform initial lock with ID and increment reentrant count
   - If there is a lock and the ID matches, increment the reentrant count
   - If there is a lock and the ID does not match, wait for the lock
1. Call the actor ensuring to attach the reentrancy ID as a header: `dapr-reentrancy-id`
1. Decrement the reentrant count and unlock if the count is 0

### SDKs
The primary responsibilities of the SDKs are to:

- Provide the reentrant specific config when an actor application starts
- Passthrough the reentrant headers

#### Providing the Configuration
This will vary based on the SDKs implementing language, but each SDK must modify their means of providing the configuration to include these new values. This must be done in whatever means they've chosen to implement the `dapr/config` API for an Actor. For instance, the Java SDK does this via annotations and could be modified like this:

```java
public interface MyActor {
  @ActorMethod(name = "say", reentrant="true")
  String say(String somethingToSay);
}

@ActorType(name = "MyActorTest", reentrant="true")
public class MyActorImpl extends AbstractActor implements MyActor, Remindable<String> {

  @Override
  public String say(String something) {
    String reversedString = "";
    try {
      this.formatAndLog(true, "say");
      reversedString = new StringBuilder(something).reverse().toString();

      this.formatAndLog(false, "say");
    } catch(Exception e) {
      // We don't throw, but the proxy side will know it failed because it expects a reversed input
      System.out.println("Caught " + e);
    }

    return reversedString;
  }
}
```

#### Passing Headers
The SDKs must also maintain any headers involving reentrancy. These headers are used to keep the request associated with the same reentrant chain and without them could not bypass any locks. Currently, there is only a single header that must be maintained: `dapr-reentrant-id`.

**Comment by halspang** [comment]
Note for a decision that was made during the PR process. We've decided to no longer support the `methods` field in the reentrancy config. Upon consideration it seemed better as an incremental improvement as we can determine what level of control users may want.

Actor config is now:

```json
{
  "entities":["actorType1", "actorType2"],
  "actorIdleTimeout": "1h",
  "actorScanInterval": "30s",
  "drainOngoingCallTimeout": "30s",
  "drainRebalancedActors": true,
  "reentrancy": {
    "enabled": true,
    "limit": 32
  }
}
```

**Comment by berndverst** [comment]
Just in case anyone ends up here, the `limit` param was renamed to `maxStackDepth`.

```
{
  "entities":["actorType1", "actorType2"],
  "actorIdleTimeout": "1h",
  "actorScanInterval": "30s",
  "drainOngoingCallTimeout": "30s",
  "drainRebalancedActors": true,
  "reentrancy": {
    "enabled": true,
    "maxStackDepth": 32
  }
}
```
